### DISCUSSION TITLE
How does the object referencing in Imports work?

### DISCUSSION THREAD

**Discussion by Bloodiko** [netbox-community/netbox#6718] [Q&A] [answered]
Hello, 

i've searched far and wide on how to reference objects correctly. 

Ive seen multiple issues regarding this, but not an correct statement in those on how its done, neither have i found something in the documentation. 

Thus here: 

I've been trying to Import devices - but the parent device does not have an name - thus the question: 
How can I reference the device ID ( Primary Key ? ) in my import correctly ? 

Thanks

**Comment by candlerb** [comment]
What kind of import: CSV? REST API / pynetbox? Django API / nbshell?

Have you tried using the device ID?  Did it work or did you get an error?  If so, show the data you submitted and what error you got.

A parent device is normally a blade chassis or some similarly important piece of kit, and therefore it usually makes sense to give it a name and to label it.  Otherwise, you have to refer to it as "that HP blade chassis that sits in rack X position Y".

**Reply by Bloodiko** [reply]
It's a CSV Import - 

in the Parent coloumn I tried to enter
{{device.id=1234}}
and many similiar tries like parent.pk=
but still object not found

Import Line (Generified, non public data) 
-----------------------------------------------------------------------------------
newDeviceName,Dell Servers,DellEMC,FC630,AssetTag,active,Slot 2,cluster,7570,Slot 2
-----------------------------------------------------------------------------------^^--------
                                                                                                                      DeviceID
Object not found
Device_Bay not found 

(Device with ID 7570 does have this Bay)

**Reply by candlerb** [reply]
I have no idea where you got that double-bracket syntax from.  This isn't a Jinja template.

Can you show what *actual* CSV data you provided, including the header line?  I am running the latest netbox v2.11.8, and the device CSV upload (`/dcim/devices/import/`) does not offer "parent" as one of the columns.  If I try to upload a CSV with a parent column, like this:

```
device_role,manufacturer,device_type,status,site,parent
Server,Unknown,Unknown,active,Test Site,1
```

 it simply says:

> Unexpected column header "parent" found.

**Reply by Bloodiko** [reply]
this double bracket i got from the issue, in which was commented by the maintainer that it should work this way - but it has been implemented differently  
and I dont know how 

I am not importing Devices - I am Importing "Child Devices" in a Housing - thus you cannot find the Parent in the normal import 

There is a seperate import at /dcim/devices/import/child-devices/ 
There you can find the mentioned "Parent" Column 

Also , i am not pasting the actual line here - its confidential data - not pasting those. 

```
name,device_role,manufacturer,device_type,asset_tag,status,comments,cluster,parent,device_bay
newDeviceName,Dell Servers,DellEMC,FC630,AssetTag,active,Slot 2,cluster,7570,Slot 2
```

the field parent is accessed by its name - and i'm trying to figure out how i can tell him to access it with the ID instead.

**Reply by candlerb** [reply]
> this double bracket i got from the issue, in which was commented by the maintainer that it should work this way - but it has been implemented differently

Which issue? Please link to your source.  I seem to remember reading somewhere that some columns support JSON, which would be `{"id":7570}`, but I can't find a reference to that currently.

> I am not importing Devices - I am Importing "Child Devices" in a Housing - thus you cannot find the Parent in the normal import
> 
> There is a seperate import at /dcim/devices/import/child-devices/

OK, I see there's a tab for that now, thanks.

> ```
> name,device_role,manufacturer,device_type,asset_tag,status,comments,cluster,parent,device_bay
> newDeviceName,Dell Servers,DellEMC,FC630,AssetTag,active,Slot 2,cluster,7570,Slot 2
> ```
> 
> the field parent is accessed by its name - and i'm trying to figure out how i can tell him to access it with the ID instead.

That's a good enough example.  I can now replicate by creating device types "Test Parent" and "Test Child", an unnamed parent device (id 72), adding device bays to device 72 (not to the device type - I hit #6714), and then trying to import a child:

```
device_role,manufacturer,device_type,status,parent,device_bay
Test,Unknown,Test Child,active,72,bay 1
```

```
device_role,manufacturer,device_type,status,parent,device_bay
Test,Unknown,Test Child,active,{"id":72},bay 1
```

Neither of these find the parent.

It seems to me that you should name your parent devices.

**Reply by candlerb** [reply]
Found it in [Netbox v2.8.2 release notes](https://github.com/netbox-community/netbox/releases/v2.8.2):

> #3147 - Allow specifying related objects by arbitrary attribute during CSV import

(implemented in #4564).  This suggests that the parent ought to be selectable by:

```
device_role,manufacturer,device_type,status,parent,device_bay
Test,Unknown,Test Child,active,{id:72},bay 1
```

But that doesn't work for me; maybe this attribute doesn't support it.  In any case, I can't see any examples of test cases in this format in `netbox/dcim/tests/test_views.py`

**Reply by candlerb** [reply]
Aha, got it.  This works: put `parent.id` as the CSV column name.

```
device_role,manufacturer,device_type,status,parent.id,device_bay
Test,Unknown,Test Child,active,72,bay 1
```

The posting at the top of #3147 was just a thought, not an actual specification.

**Reply by Bloodiko** [reply]
Yeah - just figured it out too - that i need to set it as column name 
I'll copy paste it into a new Answer to mark it as Solved 

overriding single lines probably isnt possible , but this is quite enough

Thx  though

**Comment by Bloodiko** [comment, marked as answer]
###Answer Copy Pasted from 
https://github.com/netbox-community/netbox/discussions/6718#discussioncomment-978862
------------

Aha, got it. This works: put `parent.id` as the CSV column name.

```
device_role,manufacturer,device_type,status,parent.id,device_bay
Test,Unknown,Test Child,active,72,bay 1
```
The posting at the top of #3147 was just a thought, not an actual specification.
